### PR TITLE
Store the dispatch_list in ETS, not application:set_env

### PR DESCRIPTION
**Pull request by Vagabond**
Use a protected, named table, so everyone can read it, but only
webmachine_router can write to it. The owner of the table is
webmachine_router with the heir of the process being the supervisor.

If webmachine_router is restarted, the supervisor gives ownership of the
table to the newly spawned webmachine_router process again. This means
that routing configuration is not lost if webmachine_router crashes.

The overall goal of this is to allow webmachine routes to be unregistered in an application's prep_stop callback. Calls to applicationLset_env in prep_stop go into a message loop because the application process is busy calling prep_stop().

cc @jrwest

### CODE REVIEW

**Comment by jrwest** [comment]
ran riak against this branch and did a few basho_bench runs against it + some by hand testing. No errors, so +1

File: src/webmachine_router.erl [hunk 1]
```diff
@@ -88,10 +89,40 @@ remove_resource(Resource) when is_atom(Resource) ->
...
+init_routes(DefaultRoutes) ->
+    gen_server:call(?SERVER, {init_routes, DefaultRoutes}, infinity).
+
 %% @spec start_link() -> {ok, pid()} | {error, any()}
 %% @doc Starts the webmachine_router gen_server.
 start_link() ->
-  gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+    %% We expect to only be called from webmachine_sup
+    %%
+    %% Set up the ETS configuration table, with the heir as outselves
```
Comment (russelldb) [inline src/webmachine_router.erl]: s/outselves/ourselves

File: src/webmachine_mochiweb.erl [hunk 2]
```diff
@@ -73,7 +73,7 @@ stop() ->
 
 loop(MochiReq) ->
     Req = webmachine:new_request(mochiweb, MochiReq),
-    {ok, DispatchList} = application:get_env(webmachine, dispatch_list),
+    [{dispatch_list, DispatchList}] = ets:lookup(webmachine_router, dispatch_list),
```
Comment (russelldb) [inline src/webmachine_mochiweb.erl]: For the sake of encapsulation, why not go through webmachine router to get the dispatch list?

Comment (Vagabond) [inline src/webmachine_mochiweb.erl]: I don't know... performance?  I just ported the existing behaviour.

Comment (jrwest) [inline src/webmachine_mochiweb.erl]: or perhaps similar to https://github.com/basho/riak_core/commit/40d49f9864b5528b97db6a8e54626a7fb61785a7 make the ets call in a function defined in `webmachine_router`?

EDIT: just referring to encapsulation of the ets call in module that creates the table not the fallback to the gen_server part.

Comment (Vagabond) [inline src/webmachine_mochiweb.erl]: Yes, we should probably encapsulate at least with an API if not with the gen_server itself.

Comment (russelldb) [inline src/webmachine_mochiweb.erl]: Yes. I meant something like opening get_dispatch_list.

**Comment by Vagabond** [comment]
Addressed concerns.

**Comment by bookshelfdave** [comment]
Tested with https://github.com/basho/riak_test/pull/128, these changes seem fine. Awaiting addition changes from @Vagabond.

**Comment by Vagabond** [comment]
Changed to use a public table owned by the supervisor instead.

**Comment by bookshelfdave** [comment]
riak_test replication2 passes.
all eunit tests pass.
+1

**Pull request merged**